src/util/sock.h

```diff
@@ -120,26 +121,84 @@ class Sock
+     * exceptional condition has occurred on the socket or if it has been disconnected.
+     */
+    static constexpr Event ERR = 0b100;
 
     /**
      * Wait for readiness for input (recv) or output (send).
      * @param[in] timeout Wait this much for at least one of the requested events to occur.
      * @param[in] requested Wait for those events, bitwise-or of `RECV` and `SEND`.
      * @param[out] occurred If not nullptr and `true` is returned, then upon return this
-     * indicates which of the requested events occurred. A timeout is indicated by return
-     * value of `true` and `occurred` being set to 0.
+     * indicates which of the requested events occurred (`ERR` will be added, even if
+     * not requested, if an exceptional event occurs on the socket).
+     * A timeout is indicated by return value of `true` and `occurred` being set to 0.
      * @return true on success and false otherwise
      */
     [[nodiscard]] virtual bool Wait(std::chrono::milliseconds timeout,
                                     Event requested,
                                     Event* occurred = nullptr) const;
 
+    /**
+     * Auxiliary requested/occurred events to wait for in `WaitMany()`.
+     */
+    struct Events {
+        explicit Events(Event req) : requested{req}, occurred{0} {}
+        Event requested;
+        Event occurred;
+    };
+
+    struct Hash {
+        size_t operator()(const std::shared_ptr<const Sock>& s) const
+        {
+            return s ? s->m_socket : std::numeric_limits<SOCKET>::max();
+        }
+    };
+
+    struct Equal {
+        bool operator()(const std::shared_ptr<const Sock>& lhs,
+                        const std::shared_ptr<const Sock>& rhs) const
+        {
+            if (lhs && rhs) {
+                return lhs->m_socket == rhs->m_socket;
+            }
+            if (!lhs && !rhs) {
+                return true;
+            }
+            return false;
+        }
+    };
+
+    /**
+     * On which socket to wait for what events in `WaitMany()`.
+     * The `shared_ptr` is copied into the map to ensure that the `Sock` object
+     * is not destroyed and the underlying socket closed. If this happens
+     * shortly before or after we call `poll(2)` and a new socket gets created
+     * under the same file descriptor number then the report from `WaitMany()`
+     * will be bogus.
+     */
+    using WaitData = std::unordered_map<std::shared_ptr<const Sock>, Events, Hash, Equal>;
+
+    /**
+     * Same as `Wait()`, but wait on many sockets within the same timeout.
+     * @param[in] timeout Wait this much for at least one of the requested events to occur.
+     * @param[in,out] what Wait for the requested events on these sockets and set `occurred`
+     * to the events that actually occur.
+     * A timeout is indicated by return value of `true` and all `what[].occurred`
+     * being set to 0.
+     * @return true on success and false otherwise
+     */
```

Comment:
fc4bfdf maybe, if I understand correctly:

```diff
-     * to the events that actually occur.
-     * A timeout is indicated by return value of `true` and all `what[].occurred`
-     * being set to 0.
-     * @return true on success and false otherwise
+     * for the events that actually occurred.
+     * @return true on success (or timeout, if all `what[].occurred` are returned as 0), false otherwise
      */
```

Reply:
Done.